pkg/util/logger_windows.go

```diff
@@ -0,0 +1,117 @@
+package util
+
+import (
+	"runtime"
+	"strings"
+
+	"github.com/weaveworks/common/logging"
+
+	"github.com/go-kit/kit/log/level"
+
+	"github.com/go-kit/kit/log"
+	"github.com/weaveworks/common/server"
+	el "golang.org/x/sys/windows/svc/eventlog"
+)
+
+// Default name for the Grafana Agent under Windows
+const ServiceName = "Grafana Agent"
+
+//NewWindowsEventLogger creates a new logger that writes to the event log
+func NewWindowsEventLogger(cfg *server.Config) *Logger {
+	return newLogger(cfg, makeWindowsEventLogger)
+}
+
+func makeWindowsEventLogger(cfg *server.Config) (log.Logger, error) {
+	// Setup the log in windows events
+	err := el.InstallAsEventCreate(ServiceName, el.Error|el.Info|el.Warning)
+
+	// We expect an error of 'already exists' for subsequent runs,
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
+		return nil, err
+	}
+	il, err := el.Open(ServiceName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cleanup the handle when exits scope, this could be handled via explicit close but would need to change
+	// more upstream, and its not a huge issue if it hangs around slightly longer than it should
+	runtime.SetFinalizer(il, func(l *el.Log) {
+		l.Close()
+	})
+
+	//  These are setup to be writers for each Windows log level
+	//  Setup this way so we can utilize all the benefits of logformatter
+	infoLogger := newWinLogWrapper(cfg.LogFormat, func(p []byte) error {
+		return il.Info(1, string(p))
+	})
+	warningLogger := newWinLogWrapper(cfg.LogFormat, func(p []byte) error {
+		return il.Warning(1, string(p))
+	})
+
+	errorLogger := newWinLogWrapper(cfg.LogFormat, func(p []byte) error {
+		return il.Error(1, string(p))
+	})
+
+	wl := &winLogger{
+		internalLog:   il,
+		errorLogger:   errorLogger,
+		infoLogger:    infoLogger,
+		warningLogger: warningLogger,
+	}
+	return level.NewFilter(wl, cfg.LogLevel.Gokit), nil
+}
+
+// Looks through the key value pairs in the log for level and extract the value
+func getLevel(keyvals ...interface{}) level.Value {
+	for i := 0; i < len(keyvals); i++ {
+		if vo, ok := keyvals[i].(level.Value); ok {
+			return vo
+		}
+	}
+	return nil
+}
+
+func newWinLogWrapper(format logging.Format, write func(p []byte) error) log.Logger {
+	infoWriter := &winLogWriter{writer: write}
+	infoLogger := log.NewLogfmtLogger(infoWriter)
+	if format.String() == "json" {
+		infoLogger = log.NewJSONLogger(infoWriter)
+	}
+	return infoLogger
```

Comment:
Hmm, I don't think we need one of those for each underlying writer, we can do the same shared wrapper like we do with the filter.

Reply:
Ahhh, right. Never mind!